pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+
+	logger log.Logger
+}
+
+func NewBucketRuleStore(bucket objstore.Bucket, loadConcurrency int, logger log.Logger) *BucketRuleStore {
+	return &BucketRuleStore{
+		bucket:          bucket,
+		loadConcurrency: loadConcurrency,
+		logger:          logger,
+	}
+}
+
+func newRuleStoreUserBucketClient(userID string, bkt objstore.Bucket) objstore.Bucket {
+	return bucket.NewUserBucketClient(userID, bucket.NewPrefixedBucketClient(bkt, rulePrefix))
+}
+
+func (b *BucketRuleStore) listNamespacesForUser(ctx context.Context, user string) ([]string, error) {
+	userBucket := newRuleStoreUserBucketClient(user, b.bucket)
+
+	var namespaces []string
+	err := userBucket.Iter(ctx, "", func(namespace string) error {
+		namespace = strings.TrimSuffix(namespace, delim)
+		decodedNamespace, err := base64.URLEncoding.DecodeString(namespace)
+		if err != nil {
+			return err
+		}
+
+		namespaces = append(namespaces, string(decodedNamespace))
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return namespaces, nil
+}
+
+// If existing rule group is supplied, it is Reset and reused. If nil, new RuleGroupDesc is allocated.
+func (b *BucketRuleStore) getRuleGroup(ctx context.Context, userID, namespace, groupName string, rg *rules.RuleGroupDesc) (*rules.RuleGroupDesc, error) {
+	userBucket := newRuleStoreUserBucketClient(userID, b.bucket)
+	objectKey := generateRuleObjectKey(namespace, groupName)
+
+	reader, err := userBucket.Get(ctx, objectKey)
+	if userBucket.IsObjNotFoundErr(err) {
+		level.Debug(b.logger).Log("msg", "rule group does not exist", "name", objectKey)
+		return nil, rules.ErrGroupNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = reader.Close() }()
+
+	buf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if rg == nil {
+		rg = &rules.RuleGroupDesc{}
+	} else {
+		rg.Reset()
+	}
+
+	err = proto.Unmarshal(buf, rg)
+	if err != nil {
+		return nil, err
+	}
+
+	return rg, nil
+}
+
+func (b *BucketRuleStore) ListAllUsers(ctx context.Context) ([]string, error) {
+	var users []string
+	err := b.bucket.Iter(ctx, rulePrefix, func(user string) error {
+		user = strings.TrimSuffix(strings.TrimPrefix(user, rulePrefix+delim), delim)
+		iterErr := tenant.ValidTenantID(user)
+
+		// Ignore but log invalid users in the backend. We don't want a single corrupted tenant to break
+		// rules for everyone else.
+		if iterErr != nil {
+			level.Error(b.logger).Log("msg", "invalid tenant id found in rule store", "err", iterErr, "user", user)
+		}
+
+		users = append(users, user)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list users in rule store bucket, %w", err)
+	}
+
+	return users, nil
+}
+
+func (b *BucketRuleStore) ListAllRuleGroups(ctx context.Context) (map[string]rules.RuleGroupList, error) {
+	users, err := b.ListAllUsers(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	perUserRuleGroupListMap := map[string]rules.RuleGroupList{}
+
+	for _, user := range users {
```

Comment:
We can call `ListRuleGroupsForUserAndNamespace` for each user with empty namespace, and put results to a map. No need to reimplement it here. Also we could do it concurrently to run faster.

Reply:
Makes sense, I'll keep in sequential for now and make it concurrent in a future PR.